orbit/pkg/setup_experience/setup_experience.go

```diff
@@ -13,28 +21,226 @@ type Client interface {
 	}
 
-	// poll the status endpoint
-	_, err := s.OrbitClient.GetSetupExperienceStatus()
+	// Poll the status endpoint. This also releases the device if we're done.
+	payload, err := s.OrbitClient.GetSetupExperienceStatus()
 	if err != nil {
 		return err
 	}
 
-	// TODO: fill this in!
+	// If swiftDialog isn't up yet, then launch it
+	if err := s.startSwiftDialog(binaryPath, payload.OrgLogoURL); err != nil {
+		return err
+	}
+
+	// Defer this so that s.started is only false the first time this function runs.
+	defer func() { s.started = true }()
+
+	select {
+	case <-s.closeChan:
+		log.Debug().Str("receiver", "setup_experiencer").Msg("swiftDialog closed")
+		return nil
+	default:
+		// ok
+	}
+
+	// We're rendering the initial loading UI (shown while there are still profiles, bootstrap package,
+	// and account configuration to verify) right off the bat, so we can just no-op if any of those
+	// are not terminal
+
+	if payload.BootstrapPackage != nil && payload.BootstrapPackage.Status == fleet.MDMBootstrapPackagePending {
+		return nil
+	}
+
+	if anyProfilePending(payload.ConfigurationProfiles) {
+		return nil
+	}
+
+	if payload.AccountConfiguration != nil && payload.AccountConfiguration.Status == "pending" {
+		return nil
+	}
+
+	// Now render the UI for the software and script.
+	if len(payload.Software) > 0 || payload.Script != nil {
+		var stepsDone int
+		var prog uint
+		steps := append(payload.Software, payload.Script)
+		for _, r := range steps {
+			item := resultToListItem(r)
+			if s.started {
+				err = s.sd.UpdateListItemByTitle(item.Title, item.StatusText, item.Status)
+				if err != nil {
+					log.Info().Err(err).Msg("updating list item in setup experience UI")
+				}
+			} else {
+				err = s.sd.AddListItem(item)
+				if err != nil {
+					log.Info().Err(err).Msg("adding list item in setup experience UI")
+				}
+			}
+			if r.Status == fleet.SetupExperienceStatusFailure || r.Status == fleet.SetupExperienceStatusSuccess {
+				stepsDone++
+				// The swiftDialog progress bar is out of 100
+				for range int(float32(1) / float32(len(steps)) * 100) {
+					prog++
+				}
+			}
+		}
+
+		if err = s.sd.UpdateProgress(prog); err != nil {
+			log.Info().Err(err).Msg("updating progress bar in setup experience UI")
+		}
+
+		if err := s.sd.ShowList(); err != nil {
+			log.Info().Err(err).Msg("showing progress bar in setup experience UI")
+		}
+
+		if err := s.sd.UpdateProgressText(fmt.Sprintf("%.0f%%", float32(stepsDone)/float32(len(steps))*100)); err != nil {
+			log.Info().Err(err).Msg("updating progress text in setup experience UI")
+		}
+
+		if stepsDone == len(steps) {
+			if err := s.sd.SetMessage(doneMessage); err != nil {
+				log.Info().Err(err).Msg("setting message in setup experience UI")
+			}
+
+			if err := s.sd.CompleteProgress(); err != nil {
+				log.Info().Err(err).Msg("completing progress bar in setup experience UI")
+			}
 
+			// need to call this because SetMessage removes the list from the view for some reason :(
```

Comment:
Very odd

Reply:
yeah idk why it does this. I tested it out by manually editing the file, same thing. 🤷